### PR TITLE
Abstract contract contructors

### PR DESCRIPTION
**Pull request by LefterisJP**
Implementing last bullet point of the [Abstract contracts story](https://www.pivotaltracker.com/story/show/88344782) converning constructors of base contracts and proper checks for providing all arguments. 

If not all arguments to base constructors are provided the contract is marked as abstract.

### CODE REVIEW

File: libsolidity/AST.cpp [hunk 1]
```diff
@@ -281,7 +323,7 @@ void InheritanceSpecifier::checkTypeRequirements()
 	ContractDefinition const* base = dynamic_cast<ContractDefinition const*>(m_baseName->getReferencedDeclaration());
 	solAssert(base, "Base contract not available.");
 	TypePointers parameterTypes = ContractType(*base).getConstructorType()->getParameterTypes();
-	if (parameterTypes.size() != m_arguments.size())
+	if (m_arguments.size() != 0 && parameterTypes.size() != m_arguments.size())
```
Comment (LefterisJP) [inline libsolidity/AST.cpp]: will change to `empty()`

File: mix/CodeModel.cpp [hunk 2]
```diff
@@ -135,8 +135,8 @@ CompiledContract::CompiledContract(const dev::solidity::CompilerStack& _compiler
 
 	CollectDeclarationsVisitor visitor(&m_functions, &m_locals, &m_storage);
 	contractDefinition.accept(visitor);
-	m_assemblyItems = filterLocations(_compiler.getRuntimeAssemblyItems(name), contractDefinition, m_functions);
-	m_constructorAssemblyItems = filterLocations(_compiler.getAssemblyItems(name), contractDefinition, m_functions);
+	m_assemblyItems = filterLocations(*_compiler.getRuntimeAssemblyItems(name), contractDefinition, m_functions);
```
Comment (LefterisJP) [inline mix/CodeModel.cpp]: @arkpar There is a minor, tiny touch to Mix's codebase here so I think it's prudent to also notify you.

File: libsolidity/AST.h [hunk 3]
```diff
@@ -376,7 +377,7 @@ class EnumValue: public Declaration
 
 	virtual void accept(ASTVisitor& _visitor) override;
 	virtual void accept(ASTConstVisitor& _visitor) const override;
-	TypePointer getType(ContractDefinition const* = nullptr) const;
+	TypePointer getType(ContractDefinition const* = nullptr) const override;
```
Comment (LianaHus) [inline libsolidity/AST.h]: "virtual"?

**Comment by LefterisJP** [comment]
Note: Some end to end test are failing. Looking into it.

File: libsolidity/AST.h (hunk 3 above)
Comment (LefterisJP) [inline libsolidity/AST.h]: thanks! Will do

**Comment by LefterisJP** [comment]
rebased

File: libsolidity/AST.cpp [hunk 4]
```diff
@@ -152,6 +153,50 @@ void ContractDefinition::checkAbstractFunctions()
...
+		{
+			ContractDefinition const* baseContract = dynamic_cast<ContractDefinition const*>(
+				base->getName()->getReferencedDeclaration());
+			solAssert(baseContract, "");
+			FunctionDefinition const* baseConstructor = baseContract->getConstructor();
+			if (argumentsNeeded.count(baseConstructor) == 1)
+				argumentsProvided.insert(baseConstructor);
+		}
+	}
+	// add this contract's constructor to the provided too
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Wouldn't it be better to remove them from "needed"?

File: libsolidity/AST.cpp [hunk 5]
```diff
@@ -152,6 +153,50 @@ void ContractDefinition::checkAbstractFunctions()
...
+			if (!constructor->getParameters().empty())
+				argumentsNeeded.insert(constructor);
+			for (auto const& modifier: constructor->getModifiers())
+			{
+				auto baseContract = dynamic_cast<ContractDefinition const*>(
+					modifier->getName()->getReferencedDeclaration());
+				if (baseContract)
+				{
+					FunctionDefinition const* baseConstructor = baseContract->getConstructor();
+					if (argumentsNeeded.count(baseConstructor) == 1)
```
Comment (chriseth) [inline libsolidity/AST.cpp]: baseConstructor might be nullptr - we should not insert that into argumentsProvided. Please add a test for that case.

Comment (chriseth) [inline libsolidity/AST.cpp]: Why this check here? There should be an assertion that we do not provide arguments unless they are needed. Also note that `baseConstructor` might be `nullptr` here.

File: libsolidity/AST.cpp [hunk 6]
```diff
@@ -152,6 +153,50 @@ void ContractDefinition::checkAbstractFunctions()
...
+			}
+		}
+
+		for (ASTPointer<InheritanceSpecifier> const& base: contract->getBaseContracts())
+		{
+			ContractDefinition const* baseContract = dynamic_cast<ContractDefinition const*>(
+				base->getName()->getReferencedDeclaration());
+			solAssert(baseContract, "");
+			FunctionDefinition const* baseConstructor = baseContract->getConstructor();
+			if (argumentsNeeded.count(baseConstructor) == 1)
```
Comment (chriseth) [inline libsolidity/AST.cpp]: Same applies here.

**Comment by chriseth** [comment]
Please also add an end to end test that uses this feature.

**Pull request merged**